external/storm-jms/src/main/java/org/apache/storm/jms/spout/JmsSpout.java

```diff
@@ -18,164 +18,124 @@
-    /**
-     * Stores incoming messages for later sending.
-     */
-    private LinkedBlockingQueue<Message> queue;
-    /**
-     * Contains all message ids of messages that were not yet acked.
-     */
-    private TreeSet<JmsMessageID> toCommit;
-    /**
-     * Maps between message ids of not-yet acked messages, and the messages.
-     */
-    private HashMap<JmsMessageID, Message> pendingMessages;
-    /**
-     * Counter of handled messages.
-     */
+
+    /** Counter of handled messages. */
     private long messageSequence = 0;
-    /**
-     * The collector used to emit tuples.
-     */
+
+    /** The collector used to emit tuples. */
     private SpoutOutputCollector collector;
-    /**
-     * Connection to the jms queue.
-     */
+
+    /** Connection to the jms queue. */
     private transient Connection connection;
-    /**
-     * The active jms session.
-     */
+
+    /** The active jms session. */
     private transient Session session;
-    /**
-     * Indicates whether or not a message failed to be processed.
-     */
-    private boolean hasFailures = false;
-    /**
-     * Schedules recovery tasks periodically.
-     */
-    private Timer recoveryTimer = null;
 
     /**
-     * Time to wait between recovery attempts.
+     * The message consumer.
      */
-    private long recoveryPeriodMs = -1; // default to disabled
+    private MessageConsumer consumer;
 
     /**
-     * Translate the {@code int} value of an acknowledgment to a {@code String}.
+     * Sets the JMS Session acknowledgement mode for the JMS session.
      *
-     * @param deliveryMode the mode to translate.
-     * @return its {@code String} explanation (name).
+     * <p>Possible values:
+     * <ul>
+     * <li>javax.jms.Session.AUTO_ACKNOWLEDGE</li>
+     * <li>javax.jms.Session.CLIENT_ACKNOWLEDGE</li>
+     * <li>javax.jms.Session.DUPS_OK_ACKNOWLEDGE</li>
+     * </ul>
      *
-     * @see Session
+     * Any other vendor specific modes are not supported.
+     *
+     * @param mode JMS Session Acknowledgement mode
      */
-    private static String toDeliveryModeString(int deliveryMode) {
-        switch (deliveryMode) {
+    public void setJmsAcknowledgeMode(final int mode) {
+        switch (mode) {
             case Session.AUTO_ACKNOWLEDGE:
-                return "AUTO_ACKNOWLEDGE";
-            case Session.CLIENT_ACKNOWLEDGE:
-                return "CLIENT_ACKNOWLEDGE";
             case Session.DUPS_OK_ACKNOWLEDGE:
-                return "DUPS_OK_ACKNOWLEDGE";
+                messageHandler = new MessageHandler();
+                break;
+            case Session.CLIENT_ACKNOWLEDGE:
+                messageHandler = new ClientAckHandler();
+                break;
+            case Session.SESSION_TRANSACTED:
+                messageHandler = new TransactedSessionMessageHandler();
+                break;
             default:
-                return "UNKNOWN";
-
+                LOG.warn("Unsupported Acknowledge mode: "
```

Comment:
So we are allowing non-JMS standard mode to parameter, right? I agree this is needed to support specific JMS implementation which **extends** session acknowledge mode.

Reply:
Yes. Its the only way to support provider extensions.